[로또 미션] 이승엽 미션 제출합니다.

SEUNGBOONG opened the pull request:
1단계 로또 구현 5차 수정 5/22
전체적인 수정 및 하나의 메서드에서 하나의 기능을 하게 구현했습니다. 
DTO를 적용해봤는데 맞게 적용한지 모르겠습니다. 

1단계 로또 구현 4차 수정 5/21

지금 현재상태에서 도메인에서  수익률을 계산하도록 해야하는데 아웃풋뷰에서 하는것 같습니다 수정하겠습니다. 


지금 아웃풋 뷰에서 도메인을 의존하는 이슈 발생했습니다. 
그것을 해결하기 위해 [dto](https://tristy.tistory.com/54)
를 보고 공부해서 다시 푸쉬하겠습니다. 
네이밍을 전체적으로 수정했습니다. 
네이밍에 대한 아직 이해도가 부족한것 같습니다. 그래도 많이 이해하고 넘어간 것 같습니다. 

일급컬렉션에 대한 이해도가 높아진 것 같습니다. 

메서드를 추가로 구현하면서 , 쉽게 적용되는 것을 보고, 일급컬렉션에 사용 이유에 대해 이해도가 한층 성장한것 같습니다. 
생성자나 번호 생성하는 메서드에 메시지로 interface로 전달하는 방식으로 바꿨습니다. 그리고 
각각의 책임을 전부 나눠 메서드별 하나의 책임을 지게했습니다.  이것을 
[SOLID OOP단일책임원칙](https://blog.itcode.dev/posts/2021/08/13/single-responsibility-principle)을 통해 공부했습니다. 
솔리드 법칙을  정보처리기사에서 접하게되었는데 외우기힘들었는데 자바를 하면서 자연스럽게 익히게 되었습니다.
책임에 맞게 객체를 분리하고 컨트롤러에서 필요할떄 객체에 반환값을 받아 다른 객체에게 메시지를 전달해주는 방식으로 변경했는데 올바른 방식인지 확인 부탁드립니다 .














1단계 로또 구현 3차 수정했습니다. 5/20


1. 돈을 입력했을 때, 로또 수량을 계산하는 클래스를 LottoTicket 클래스에서 구현을 했고, 일급컬렉션으로 관리하기 위해서 

컬렉션 외 멤버 변수를 가지지 않게 신경썼고,  불변성 유지 및  직접 노출을 막고자 private으로 선언을 해야하나를 고민중에 있다.

저기 부분을 private 한다고해서 직접 노출을 막을 수있는지에 대해 공부하고 질문해봐야겠다. 

 
2. 로또 수량이 매개변수로 넘어오면 로또를 생성하고 관리하는 일급 컬렉션

NumberGenerator에서 인터페이스로 리스트를 선언했고, RandomNumberGenerator에서 오버라이딩해 로또 숫자 리스트를 구현했다. 이것도 일급 컬렉션으로 관리하기 위해 주요 로직은 private으로 처리하고  위에서 노출을 삼가한다는 것은 주요 로직만 private 처리를 해야한다는 것이 었나보다 라는 추측에 있습니다. 

랜덤 로또 번호만을 관리하도록 일급컬렉션을 구현해봤습니다.

3. 일급 컬렉션 안에서 관리당하는 로또 객체

LottoTicketStore에서는 일급 컬렉션 속에서 로또 객체를 관리하도록 진행을 했다. 
[공부내용 정리 수정 전](https://yeopseung.tistory.com/117)
[공부내용 수정 후](https://yeopseung.tistory.com/122)


















1단계 로또 구현 2차 수정했습니다. 5/17 
전체적으로 리팩토링 했습니다.
일급컬렉션을 통한 데이터를 관리하게 만들었습니다. 
전체적인 클래스 재정비했습니다. 
일급 컬렉션 찾아가며 공부했는데 
아직 일급 컬렉션에 대한 공부가 덜 된것이 느껴집니다. 
조금 더 공부해보겠습니다. 
[일급컬렉션 공부](https://blog.naver.com/sosow0212/223014652755?trackingCode=blog_bloghome_searchlist)
참고 하고 했습니다. 감사합니다. 













1단계 로또 구현  1차 수정했습니다. 5/15 

같은 숫자만 나오는 이슈 해결했습니다.
클래스 컨벤션 수정했습니다.
[TDD](https://mangkyu.tistory.com/182)  [인터페이스](https://interconnection.tistory.com/129)
두 가지를 통해 공부 후 최대한 적용해보려고 했습니다. 
멘토님이 말씀하시는 일급리스트가  잘 적용 됐는지 궁금합니다. 







1단계 구현했습니다.  5/14
클린 코드 MVC 모델 , 컨벤션 최대한 신경쓰고 했습니다.
테스트는 아직 미숙한게 보여 보완하겠습니다.

SEUNGBOONG committed:
feat:
- 금액 입력 메서드 구현

SEUNGBOONG committed:
feat:
- LottoNum  금액에 따라 로또 개수 추출  메서드 구현

SEUNGBOONG committed:
feat:
- Lotto 개수 추출 메서드 테스트 성공

SEUNGBOONG committed:
feat:
- Lotto 난수 추출 메서드

SEUNGBOONG committed:
feat:
- 금액 입력 메서드 구현

SEUNGBOONG committed:
feat:
- 로또 구매 개수 메서드 구현

SEUNGBOONG committed:
feat:
- 로또 난수 생성 메서드 구현
- 중복 제거 메서드 구현

SEUNGBOONG committed:
feat:
- 로또 난수 범위 테스트 성공

SEUNGBOONG committed:
Text:
- README.md 요구사항명세서 작성 완료

SEUNGBOONG committed:
Refact:
- 작동 구현

SEUNGBOONG committed:
feat:
- 출력 반복문 완성

SEUNGBOONG committed:
docs:
- 요구사항 완료

sosow0212 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.List;
+
+public class LottoController {
+    public static void main(String[] args) {
```
1. 첫 줄 개행 추가해주세요!
2. main 메서드가 static일 필요가 있을까요?! Main 클래스 하나 만들어서 Controller에 주입 및 호출하여 실행시켜보는 건 어떨까요?

SEUNGBOONG replied:
수정 했습니다. 제가 한 방식이 맞는지 확인 부탁드립니다!

sosow0212 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,16 @@
+
+public class OutputView {
+
+    private int i;
```
for문 안에서 선언해도 될 것 같습니다~

SEUNGBOONG replied:
넵 !

sosow0212 commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,14 @@
+
+    private static final int LOTTO_NUM = 1000;
+
+    private int lottoNum(int money) {
```
1. 해당 메서드가 `getLottoNum`보다 밑으로 가야할 것 같습니다.
2. 메서드 동사명으로 시작해주세요~
3. `getLottoNum`도 정확히 LottoNum이 무엇인지 헷갈리는데 네이밍 조금 더 보기 좋게 바꾸면 좋을 것 같습니다!

SEUNGBOONG replied:
수정 완료했습니다. 네이밍 선택할 때 신경쓰겠습니다.

sosow0212 commented on src/main/java/domain/LottoRandom.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.Random;
+
+public class LottoRandom {
+    private static final int LOTTO_RANGE = 45;
```
첫줄에 개행 하나 추가해주세요!

SEUNGBOONG replied:
수정 했습니다!

sosow0212 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,22 @@
+
+public class LottoController {
+    public static void main(String[] args) {
+        InputView inputView = new InputView();
+        OutputView outputView = new OutputView();
+        LottoRandom lottoRandom = new LottoRandom();
```
객체로 가지고 있는 것과 메서드 내에서 객체를 선언하는 것과 무엇이 다를까요?

SEUNGBOONG replied:
1. 객체를 가지고 있는 것은  객체에 접근하고 조작할 수 있고 , 필요에 따라 여러 곳에서 참조할 때 사용될 것 같습니다.
2. 메서드 내에서 객체를 선언하는 것 : 메서드가 실행 될떄마다 새로운 객체를 생성해야 할 때 사용되는 것 같습니다. 혹시 추천해주실 블로그 있을까요 ?

sosow0212 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,22 @@
+
+import java.util.List;
+
+public class LottoController {
```
<img width="456" alt="image" src="https://github.com/next-step/java-lotto-clean-playground/assets/63213487/4c849189-2c03-42d0-92b1-a9edba657e7b">

돌려봤는데 로또 번호가 모두 똑같이 나오고, 2000원이면 2장이 나와야하는데 3장이 나오네요🥲
lottoPaper을 콘솔에 찍어보니 객체를 생성하기 전에 이미 생성되는 걸 보니 로직이 조금 잘못된 것 같아요!
이 부분 수정해주세요~

sosow0212 commented on src/main/java/domain/LottoRandom.java:
```diff
@@ -0,0 +1,32 @@
+        List<Integer> lottoNumbers = new ArrayList<>();
+        while (lottoNumbers.size() <= LOTTO_COUNT) {
+            int randomNumber = random.nextInt(LOTTO_RANGE) + LOTTO_NUM;
+            duplicationRemove(lottoNumbers, randomNumber);
```
```suggestion
            removeDuplicationNumber(lottoNumbers, randomNumber);
```

sosow0212 commented on src/main/java/domain/LottoRandom.java:
```diff
@@ -0,0 +1,32 @@
+
+    private List<Integer> lottoRandomNum() {
+        List<Integer> lottoNumbers = new ArrayList<>();
+        while (lottoNumbers.size() <= LOTTO_COUNT) {
```
알고리즘 슬슬 푸시더니 while로 잘 구현하셨군요ㅋㅋㅎㅎ 👍

sosow0212 commented on src/main/java/domain/LottoRandom.java:
```diff
@@ -0,0 +1,32 @@
+        return lottoNumbers;
+    }
+
+    private static void duplicationRemove(List<Integer> lottoNumbers, int randomNumber) {
+        if (!lottoNumbers.contains(randomNumber)) {
+            lottoNumbers.add(randomNumber);
+        }
+    }
```
1. 해당 메서드가 static일 필요가 없어보입니다!
2. `포함되지 않으면 숫자를 삽입한다` 라는 기능을 수행하는 것 같은데 putNumbersIfNotContains 이런 네이밍이나 더 와닿는 걸로 메서드명 변경하면 좋을 것 같습니다!

sosow0212 commented on src/main/java/domain/LottoRandom.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.List;
+import java.util.Random;
+
+public class LottoRandom {
```
로또의 랜덤 번호를 생성하는 건 어떤 객체의 역할일까요?

`일급 컬렉션`과 `원시값 포장` 키워드에 대해 학습하고 적용해보면 더 좋을 것 같아요!

sosow0212 commented on src/test/java/domain/LottoRandomTest.java:
```diff
@@ -0,0 +1,22 @@
+        }
+    }
+
+}
```
EOF 에러가 아직 있습니다!

sosow0212 commented on src/test/java/domain/LottoRandomTest.java:
```diff
@@ -0,0 +1,22 @@
+class LottoRandomTest {
+    @Test
+    void 로또_난수_범위_테스트() {
+        // given
+        LottoRandom lottoRandom = new LottoRandom();
+        List<Integer> lottoNum=lottoRandom.getLottoRandomNum();
+        // when&then
+        for (int num : lottoNum) {
+            assertTrue(1 <= num && num <= 45,"범위오버");
+        }
+    }
```
```suggestion
        // given
        LottoRandom lottoRandom = new LottoRandom();
        
        // when 
        List<Integer> lottoNum=lottoRandom.getLottoRandomNum();
        
        // then
        for (int num : lottoNum) {
            assertTrue(1 <= num && num <= 45,"범위오버");
        }
    }
```

1. 가독성을 위해서 이렇게 바꾸면 더 좋을 것 같아요!
2. 정렬이 안되어 있습니다
3. 테스트가 잘못된 것 같습니다. `범위오버`라는 메시지를 바꾸면 테스트가 실패해야할 것 같은데 실패하지 않습니다! 다른 메서드를 활용해보면 어떨까요?

sosow0212 commented on src/test/java/domain/LottoTest.java:
```diff
@@ -0,0 +1,20 @@
+
+
+class LottoTest {
+    @Test
```
1. 첫줄 개행 추가해주세요!
2. 정렬 및 위에처럼 given-when-then 사이 개행 한 칸씩만 추가해주세요~

SEUNGBOONG replied:
넵 다음 부턴 주의하겠습니다.

sosow0212 commented on src/test/java/domain/LottoTest.java:
```diff
@@ -0,0 +1,20 @@
+        // then
+        Assertions.assertEquals(expectedNum,value );
+    }
+
+
```
불필요한 개행 있습니다 + EOF 해결해주세요!

SEUNGBOONG replied:
넵

sosow0212 commented on src/test/java/domain/LottoTest.java:
```diff
@@ -0,0 +1,20 @@
+        int value = lotto.getLottoNum(14000);
+        int expectedNum = 14;
+        // then
+        Assertions.assertEquals(expectedNum,value );
```
검증 좋습니다 👍

SEUNGBOONG replied:
감사합니다!!!!!!

sosow0212 commented on src/test/java/domain/LottoRandomTest.java:
```diff
@@ -0,0 +1,22 @@
+
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
```
import에 와일드카드는 지양해주세요! 왜 그럴까요?!
[설정 방법](https://yanoo.tistory.com/152)

SEUNGBOONG replied:
보내주신 URL 을 통해 수정했습니다 감사합니다.

sosow0212 commented on src/main/java/domain/Lotto.java:
```diff
@@ -0,0 +1,14 @@
+package domain;
+
+public class Lotto {
```
현재 로또 객체에 맞는 일을 하지 못하고 있는 것 같아요!

SEUNGBOONG replied:
수정완료했습니다.

sosow0212 reviewed (changes requested):
굉장히 빠르게 만드셨네요 👏

실행해보니 결과가 잘 안나오는 것 같아요!
구조적으로 문제가 있어보이는데 이 부분 고민하고 해결해보시면 좋을 것 같습니다.

원시값 포장, 일급 컬렉션 키워드에 대해서 학습해봐도 좋고, 현재 필요한 객체는 무엇이고 객체가 해야할 일은 무엇일지 생각해보면 도움될 것 같습니다!

SEUNGBOONG committed:
Fix:
CountLotto로 클래스 이름 변경

SEUNGBOONG committed:
Fix:
-LottoService로 클래스 변경

Refactor:
- RandomNumber 인터페이스 생성후 오버라이드 재사용

SEUNGBOONG committed:
Fix :
- OutputView 에서 사용하던 반복문 Controller로 재구현했습니다.

SEUNGBOONG committed:
Fix :
- 랜덤 숫자를 테스트 하기위한 가상 로또 컨트롤러 입니다.

SEUNGBOONG committed:
Feat :
- 로또 테스트를 위한 랜덤 고정값 클래스 입니다.

SEUNGBOONG committed:
Feat :
- 인터페이스에서 생성 후 오버라이드를 통해 구현하도록 진행했습니다.

SEUNGBOONG committed:
Feat :
- TDD 방식을 읽고
테스트 주도 개발 방식을 생각해 무작위 숫자를 테스트 할 수없다를 판단해 인터페이스를 구현 후 오버라이딩 했습니다.

SEUNGBOONG committed:
Test :
- 로또 난수 범위가 1~45 까지 인지 확인 했습니다.

- 숫자를 고정 후 알맞게 잘 진행 되는지 테스트

SEUNGBOONG committed:
Test :
- 로또 난수 범위가 1~45 까지 인지 확인 했습니다.

- 숫자를 고정 후 알맞게 잘 진행 되는지 테스트

SEUNGBOONG committed:
Fix:
-Stream으로 수정

SEUNGBOONG committed:
Fix:
-개행 추가

SEUNGBOONG committed:
Fix:
-개행 추가

SEUNGBOONG committed:
Fix:
-개행 추가

SEUNGBOONG committed:
Fix:
-개행 추가

SEUNGBOONG committed:
Fix:
- Stream 말고 다른 방법 사용

SEUNGBOONG committed:
Fix:
- 컨트롤러에 주입 및 호출하여 실행 하는 방식으로 수정 했습니다.

sosow0212 commented on src/main/java/domain/FakeGenerateRandomNumber.java:
```diff
@@ -0,0 +1,13 @@
+package domain;
```
테스트에서 사용하면 테스트 패키지로 옮겨주세요~

SEUNGBOONG replied:
완료했습니다 .

sosow0212 commented on src/main/java/domain/GenerateRandomNumber.java:
```diff
@@ -0,0 +1,33 @@
+import java.util.Random;
+
+public class GenerateRandomNumber implements RandomNumber {
+    private static final Random random = new Random();
```
첫줄 개행이 빠졌습니다!

SEUNGBOONG replied:
넵 감사합니다

sosow0212 commented on src/main/java/domain/GenerateRandomNumber.java:
```diff
@@ -0,0 +1,33 @@
+            lottoNumber.add(randomNumber);
+        }
+    }
+
```
불필요한 개행이 있습니다

SEUNGBOONG replied:
개행이 아직 미숙한 것 같습니다 조금 더 신경써보도록하겠습니다.

sosow0212 commented on src/main/java/domain/GenerateRandomNumber.java:
```diff
@@ -0,0 +1,33 @@
+        return generatedNumber(lottoNumber);
+    }
+
+    private static List<Integer> generatedNumber(List<Integer> lottoNumber) {
```
이거랑 아래 메서든느 static일 필요가 없을 것 같습니다~

SEUNGBOONG replied:
넵 수정했습니다.

SEUNGBOONG replied:
개념을 확실히 정리했다 생각했는데 한번 더 Static에 대해 집고 넘어가보겠습니다.

sosow0212 commented on src/main/java/domain/GenerateRandomNumber.java:
```diff
@@ -0,0 +1,33 @@
+        return lottoNumber;
+    }
+
+    private static void generatedAddNumber(List<Integer> lottoNumber, int randomNumber) {
```
메서드는 동사로 시작해주세요!
의미에 맞게 -> `removeIfContainsDuplicatedNumber(...)` 요런 느낌의 네이밍은 어떨까요?

SEUNGBOONG replied:
넵 수정완료했습니다.

SEUNGBOONG replied:
마침 네이밍 고민이 있었습니다!

sosow0212 commented on src/main/java/controller/Main.java:
```diff
@@ -0,0 +1,28 @@
+        CountLotto countLotto = new CountLotto();
+        GenerateRandomNumber generateRandomNumber = new GenerateRandomNumber();
+
+        int money = inputView.getMoney();
+        int paper = countLotto.getCountLotto(money);
+
+        for (int i = INITIAL_NUM; i < paper; i++) {
+            List<Integer> lottoList = generateRandomNumber.generateNumber();
+            outputView.lotto(lottoList);
+        }
```
main은 실행부로 두고
컨트롤러에서 input, output 을 담당하는게 MVC패턴에 적절할 것 같습니다.
랜덤 번호 생성도 객체로 빼는 건 어떨까요?

sosow0212 commented on src/main/java/domain/CountLotto.java:
```diff
@@ -0,0 +1,14 @@
+package domain;
+
+public class CountLotto {
```
해당 객체는 로또 몇 장인지를 나타내는 것 같은데 네이밍을 변경하면 좋을 것 같습니다 지금 네이밍은 처음에 알기 쉽지 않을 것 같아요! Ticket이라든 다른 거든 좋을 것 같아요!

SEUNGBOONG replied:
수정했습니다!

sosow0212 commented on src/main/java/domain/CountLotto.java:
```diff
@@ -0,0 +1,14 @@
+
+    private static final int LOTTO_ADD_PRICE = 1000;
+
+    public int getCountLotto(int money) {
+        return countLotto(money);
+    }
+
+    private int countLotto(int money) {
+        return money / LOTTO_ADD_PRICE;
+    }
```
이건 메서드 분리하지 않아도 getCountLotto의 네이밍만 신경쓰면 의미 전달이 될 것 같습니다 !

sosow0212 commented on src/test/java/domain/CountLottoTest.java:
```diff
@@ -0,0 +1,18 @@
+        // then
+        Assertions.assertEquals(expectedNum, value);
+    }
+}
```
EOF 발생하네요 ㅠ 다른 테스트에도 나타나는 것 같습니다

SEUNGBOONG replied:
EOF는 End of File의 약자로, 더 이상 읽을 수 있는 데이터가 없음을 나타내는 에러이다. 그래서 더 이상 읽을 수 있는 데이터가 없을땐 끝내줘야하는데 끝내주지 못하는 경우에 많이 등장한다. 라고 공부했는데  결국엔 무한루프인For문에서 일어나느 현상일까요>?

sosow0212 commented on src/test/java/domain/CountLottoTest.java:
```diff
@@ -0,0 +1,18 @@
+
+
+class CountLottoTest {
+    @Test
```
첫줄 개행도 지켜주세요

sosow0212 reviewed (changes requested):
오프라인에서 말한 것처럼 객체 분리랑 클래스 정리 한 번만 하고 다시 리뷰하도록 하겠습니다!

SEUNGBOONG committed:
Fix:
- Test로 이동했습니다.

SEUNGBOONG committed:
Fix:
- 개행 추가했습니다.

SEUNGBOONG committed:
Fix:
- 전체적으로 클래스를 다듬어봤는데, 어떤점이 부족한지 감이 오질않고있습니다!

SEUNGBOONG committed:
Fix:
- 불필요한 개행 제거

SEUNGBOONG committed:
Fix:
-  LottoRange-> MAX
- Lotto_NUM -> MIN

SEUNGBOONG committed:
Fix:
- LOTTO_ADD_PRICE -> TICKET_PRICE

SEUNGBOONG committed:
Refactor:
- 일급 컬렉션 및 원자 값 포장 전 코드

SEUNGBOONG committed:
Refactor:
- 일급 컬렉션을 사용한 리팩토링

SEUNGBOONG committed:
Refactor:
- 일급 컬렉션을 사용한 리팩토링

SEUNGBOONG committed:
Refactor:
- 로또 넘버 테스트를 위한 인터페이스 생성

SEUNGBOONG committed:
Refactor:
- 출력 값 가공 후 출력

SEUNGBOONG committed:
Refactor:
- 일급 컬렉션은 한 메서드만 존재해야해서 새로운 클래스 생성후 로또를 생성하는 메서드 구현

SEUNGBOONG committed:
Refactor:
- 로또 테스트를 위한 가상 로또 값생성

SEUNGBOONG committed:
Refactor:
- 로또 가상 값 테스트 및 범위 테스트

SEUNGBOONG committed:
Refactor:
- 컨트롤러에서 작동 구현

sosow0212 commented on src/main/java/domain/LottoTicket.java:
```diff
@@ -0,0 +1,29 @@
+    private final List<CreateLottoTicket> lottoTicket;
+    private static final int INITIAL_NUMBER = 0;
+    private static final int TICKET_PRICE = 1000;
+    private final int ticketCount;
```
```suggestion
    private static final int INITIAL_NUMBER = 0;
    private static final int TICKET_PRICE = 1000;
    
    private final List<CreateLottoTicket> lottoTicket;
    private final int ticketCount;
```

sosow0212 commented on src/main/java/domain/LottoTicket.java:
```diff
@@ -0,0 +1,29 @@
+        this.lottoTicket = generateLottoTickets();
+    }
+
+    public List<CreateLottoTicket> getLottoTicket() {
```
내부에서만 사용한다면 접근제어자는 private로 변경해주세요~

SEUNGBOONG replied:
변경했습니다

sosow0212 commented on src/main/java/domain/LottoTicket.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    public List<CreateLottoTicket> getLottoTicket() {
+        return new ArrayList<>(lottoTicket);
```
혹시 방어적 복사를 위해 ArrayList로 감싼 것인가요?!

SEUNGBOONG replied:
[방어적복사](https://dev-cool.tistory.com/25) 이거말씀하시는 것일까요?

sosow0212 commented on src/main/java/domain/CreateLottoTicket.java:
```diff
@@ -0,0 +1,36 @@
+import java.util.List;
+import java.util.Random;
+
+public class CreateLottoTicket implements RandomNumber {
```
1. 여기서는 로또에 관한 클래스 같은데 Create라는 네이밍이 붙으니 뭔가 생성에 관한 클래스 같습니다!
2. RandomNumber 자체를 구현한다면 CreateLottoTicket도 Number 생성에 대한 클래스인 것 같습니다. implements로 인터페이스를 구현하지말고, 생성자나 번호 생성하는 메서드에 메시지로 Interface를 전달하는 건 어떨까요?

SEUNGBOONG replied:
공부해서 해보겠습니다.. 메시지로 전달하는 것 까지 일급컬렉션의 일부일까요?

sosow0212 commented on src/main/java/domain/CreateLottoTicket.java:
```diff
@@ -0,0 +1,36 @@
+
+    private List<Integer> lottoNumber;
+
+    private static final Random random = new Random();
```
```suggestion
    private static final int MAX = 45;
    private static final int COUNT_NUMBER = 1;
    private static final int LOTTO_SIZE = 6;

    private final Random random = new Random();
    
    private List<Integer> lottoNumber;
```

sosow0212 commented on src/main/java/domain/CreateLottoTicket.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    @Override
+    public List<Integer> generateRandomNumber() {
```
해당 메서드도 내부에서만 사용한다면 private로 닫아주세요!

sosow0212 commented on src/main/java/domain/CreateLottoTicket.java:
```diff
@@ -0,0 +1,36 @@
+            int randomNumber = random.nextInt(MAX) + COUNT_NUMBER;
+            if (!number.contains(randomNumber)) {
+                number.add(randomNumber);
+            }
```
해당 부분을 묶어서 메서드 분리 및 뜻을 부여해주는 건 어떨까요?

sosow0212 commented on src/main/java/domain/LottoTicket.java:
```diff
@@ -0,0 +1,29 @@
+
+public class LottoTicket {
+
+    private final List<CreateLottoTicket> lottoTicket;
```
List<CreateLottoTicket> 자체를 한 번 더 일급 컬렉션으로 감싸는 건 어떨까요?

그러면 이런 장점이 있을 것 같아요
생성자로 `Lottos(int ticketCount)`를 받은 후 `일급 컬렉션 Lottos 내부의 List<CreateLottoTicket>의 요소를 모두 생성`

그리고 Lottos 입장에선 외부에서 메시지를 받아 로직을 숨길 수 있을 거 같은데 현재는 로또의 수도 책임지고, 로또들도 책임져서 단일책임원칙을 위반한 것 같습니다~

따라서 객체를 하나로 묶기보다는 분리를 한 후 적절한 메시지를 전달하여 역할을 수행하면 더 좋을 것 같습니다!

sosow0212 commented on src/main/java/domain/WinLotto.java:
```diff
@@ -0,0 +1,5 @@
+package domain;
+
+public class WinLotto {
```
요건 나중에 하려고 만드신거죠??
현재는 진행하지 않으니 제거해주세요!

sosow0212 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,14 @@
+
+public class OutputView {
+
+    public void displayLottoTickets(List<CreateLottoTicket> lottoTickets) {
```
`List<CreateLottoTicket>` 자체를 넘겨서 View가 Model을 의존하는 것 같습니다. (import 보면 domain 패키지가 딸려 들어갔어요!)

객체가 아닌 값 필드 혹은 List<Integer>나 Dto를 사용하여 의존을 끊어내는 건 어떨까요?

SEUNGBOONG replied:
넵 확인후 수정하겠습니다.

sosow0212 commented on src/test/java/domain/CreateLottoTicketTest.java:
```diff
@@ -0,0 +1,47 @@
+
+    @Test
+    void 로또_가상값_테스트() {
+
```
여기도 그렇고 다른 부분도 불필요한 개행입니다!

SEUNGBOONG replied:
[일급 컬렉션](www.naver.com)

sosow0212 commented on src/test/java/domain/CreateLottoTicketTest.java:
```diff
@@ -0,0 +1,47 @@
+    void 로또_가상값_테스트() {
+
+        // given
+        FakeGenerateRandomNumber fakeGenerateRandomNumber = new FakeGenerateRandomNumber();
+        CreateLottoTicket createLottoTicket = new CreateLottoTicket();
```
현재 테스트가 제대로 수행되지 않고 있습니다.
위에 리뷰 남긴 것처럼 Interface를 CreateLottoTicket에 넘겨주고, 테스트할 땐 Fake구현체를 넣어준다면 의도하신대로 테스트 결과가 나올 것 같습니다!

sosow0212 commented on src/test/java/domain/CreateLottoTicketTest.java:
```diff
@@ -0,0 +1,47 @@
+            assertTrue(1 <= num && num <= 45, "로또 숫자는 1~45까지");
+        }
+    }
+}
```
EOF 오류가 아직도 발생합니다 ㅠ

SEUNGBOONG replied:
혹시EOF오류가 for문을 사용하나요? EOF 오류가 발생하는지는 어떻게 판다할까여 ?

sosow0212 commented on src/main/java/domain/LottoTicket.java:
```diff
@@ -0,0 +1,29 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class LottoTicket {
```
현재는 LottoTicket에서는 몇 장을 사야하는지에 대한 로또 개수를 관리하는 책임도 있고 로또 번호를 생성하는 책임도 있어보입니다. 책임에 맞게 객체를 분리하고 컨트롤러에서 필요할 때 객체에 반환값을 받아 다른 객체에게 메시지로 전달해주는 방식으로 변경해보면 좋아보입니다!

SEUNGBOONG replied:
넵

sosow0212 reviewed (changes requested):
구조가 이전보다 더 좋아진 것 같습니다!
구조 중심적으로 몇 가지 리뷰를 남겼는데 확인해보고 수정하면 좋을 것 같습니다.
거의 다 왔으니 조금만 힘내세요 👍

SEUNGBOONG committed:
Refactor:
- 실행 할 수 있게 구현

SEUNGBOONG committed:
Refactor:
- 인터페이스 오버라이딩을 통한 로또 넘버 관리 구현

SEUNGBOONG committed:
Refactor:
- 가상 값이랑 범위 테스트

SEUNGBOONG committed:
Refactor:
- 가상 값 리스트 선언

SEUNGBOONG committed:
Refactor:
- 오버라이드를 통한 랜덤로또 숫자만 관리하는 메서드 구현

SEUNGBOONG committed:
Refactor:
-

SEUNGBOONG committed:
Refactor:
- 도메인과 의존하는 문제 해결

SEUNGBOONG committed:
Refactor:
- 삭제

SEUNGBOONG committed:
Fix:
- 일급 컬렉션 특성인 불변성 유지 직접 노출 금지를 지키기위한
public -> private 수정

SEUNGBOONG committed:
Fix:
- 일급 컬렉션 특성인 불변성 유지 직접 노출 금지를 지키기위한
public -> private 수정
을했지만, 확실하지 않아서 다시 public으로 진행

SEUNGBOONG committed:
Fix:
- 일급 컬렉션 특성인 불변성 유지 직접 노출 금지를 지키기위한
public -> private 수정
을했지만, 확실하지 않아서 다시 public으로 진행

SEUNGBOONG committed:
Feat:
- Stream 을 사용 한 당첨 번호 리스트화

SEUNGBOONG committed:
Feat:
- Stream 을 사용 한 당첨 번호 리스트화

SEUNGBOONG committed:
- 변수 선언 순서 변경

SEUNGBOONG committed:
Refactor:
- 네이밍 변경
- 일급컬렉션 특성상 멤버변수는 하나여야하는데 두개를 유지하고 있어 분리

SEUNGBOONG committed:
Refactor:
- 네이밍 변경
- 일급컬렉션 특성상 멤버변수는 하나여야하는데 두개를 유지하고 있어 분리
- TicektCount라는 클래스 생성후  따로 관리

SEUNGBOONG committed:
Refactor:
- 수익률 계산기 메서드 추가

SEUNGBOONG committed:
Refactor:
-  outputView에서 도메인을 의존하는 방식을 끊기 위한 Dto 생성 시도중

SEUNGBOONG committed:
Refactor:
- 로또티켓만을 관리하는 메서드 구현
- 출력값이 주소로 나오는 문제 해결

SEUNGBOONG committed:
Refactor:
- 인터페이스에서 오버라이딩 해서 로또만을 관리하게 끔 구현

SEUNGBOONG committed:
Refactor:
- 입력값을 받기 위한 메서드 구현

SEUNGBOONG committed:
Refactor:
- 로또 장수를 관리하기위한 클래스 생성후 일급컬렉션으로 관리

SEUNGBOONG committed:
Refactor:
- 로또 테스트 네이밍 바뀌는 문제등 해결

SEUNGBOONG committed:
Refactor:
- 로또 컨트롤러 합체

SEUNGBOONG committed:
Refactor:
- 리드미 수정

sosow0212 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,42 @@
+    public void run() {
+        int money = inputView.getMoney();
+        TicketCount ticketCount = new TicketCount(money);
+        LottoTickets lottoTickets = new LottoTickets(new RandomNumberGenerator());
+        List<LottoTicket> tickets = lottoTickets.generateLottoTickets(ticketCount.getCount());
```
이걸 하나로 묶을 수 있지 않을까요?

SEUNGBOONG replied:
넵 다음 STep 진행할때 해보곘습니다.

sosow0212 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,42 @@
+
+    public static void main(String[] args) {
+        new LottoController().run();
+    }
```
이것도 상위 Application 같은 클래스에서 Controller를 작동하도록 하면 좋을 것 같아요

SEUNGBOONG replied:
넵 적용하도록하곘습니다

sosow0212 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,51 @@
+        for (LottoTicket ticket : tickets) {
+            System.out.println(ticket);
+        }
+    }
```
아직 도메인을 의존하고 있네요!

sosow0212 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,51 @@
+
+    private static List<Integer> winningsMoney() {
+        return winningsMoney;
+    }
```
1. 메서드 순서 맞춰주세요!
2. static 메서드일 필요가 있나요?

sosow0212 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,51 @@
+            System.out.printf("%d개 일치 (%d원)\n", count, winningsMoney.get(INITIAL_NUMBER_TWO));
+        } else if (count == SIX_COUNT) {
+            System.out.printf("%d개 일치 (%d원)\n", count, winningsMoney.get(INITIAL_NUMBER_THREE));
+        } else {
+            System.out.printf("%d개 일치\n", count);
+        }
```
else 사용하지 않는 것도 제약조건 아니었나요!?

sosow0212 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,51 @@
+
+    public void winningLottoRateOfResult(int payMoney) {
+        for (int i = 0; i < winningsMoney().size(); i++)
+            sum += winningsMoney().get(i);
```
```suggestion
        for (int i = 0; i < winningsMoney().size(); i++) {
            sum += winningsMoney().get(i);
       }
```

sosow0212 commented on src/test/java/domain/LottoTest.java:
```diff
@@ -0,0 +1,47 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class LottoTest {
```
해당 클래스 말고 다른 클래스들은 테스트하지 않아도 될까요?

SEUNGBOONG committed:
Refactor:
- 컨트롤러 다시 작성

SEUNGBOONG committed:
Refactor:
- OutputView에서 구현했던
수익률 및, 로또 당첨금 조회를 모델에서 재구현했습니다.

SEUNGBOONG committed:
Add:
- 당첨금 클래스 생성

SEUNGBOONG committed:
Fix:
- 개행 첨삭

SEUNGBOONG committed:
Refactor:
- Dto 적용 시도 성공
- WinningSMoney 삭제 후 LottoRate 에서 관리하도록 구현

SEUNGBOONG committed:
Refactor:
- Dto 적용 시도 성공
- WinningSMoney 삭제 후 LottoRate 에서 관리하도록 구현

SEUNGBOONG committed:
Refactor:
- Dto 적용 시도 성공
- WinningSMoney 삭제 후 LottoRate 에서 관리하도록 구현

sosow0212 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,59 @@
+        this.outputView = new OutputView();
+    }
+
+    public static void main(String[] args) {
+        new LottoController().run();
+    }
```
이 부분은 상위에 Application  클래스 하나 만들어서 위임하면 좋을 것 같습니다!

sosow0212 commented on src/main/java/domain/LottoRate.java:
```diff
@@ -0,0 +1,23 @@
+            sum += matchesMoney().get(i);
+        return (double) sum / payMoney;
+    }
+}
```
EOF 문제가 발생합니다!

sosow0212 commented on src/main/java/domain/LottoRate.java:
```diff
@@ -0,0 +1,23 @@
+    private static final int THREE_MATCHES = 5000;
+    private static final int FOUR_MATCHES = 50000;
+    private static final int FIVE_MATCHES = 150000;
+    private static final int SIX_MATCHES = 20000000;
+    private int sum = 0;
```
```suggestion
    private static final int SIX_MATCHES = 20000000;
    
    private int sum = 0;
```

sum = 0을 생성자에서 해주는 건 어떨까요??

sosow0212 commented on src/main/java/domain/LottoTicket.java:
```diff
@@ -0,0 +1,21 @@
+    public String toString() {
+        return numbers.toString();
+    }
+}
```
EOF 발생합니다!

sosow0212 commented on src/main/java/domain/LottoTickets.java:
```diff
@@ -0,0 +1,24 @@
+
+public class LottoTickets {
+
+    private static final int INITIAL_NUMBER = 0;
+    private final NumberGenerator numberGenerator;
```
```suggestion
    private static final int INITIAL_NUMBER = 0;
    
    private final NumberGenerator numberGenerator;
```

sosow0212 commented on src/main/java/domain/LottoTickets.java:
```diff
@@ -0,0 +1,24 @@
+        this.numberGenerator = numberGenerator;
+    }
+
+    public List<LottoTicket> generateLottoTickets(int ticketCount) {
+        List<LottoTicket> tickets = new ArrayList<>();
+        for (int i = INITIAL_NUMBER; i < ticketCount; i++) {
+            List<Integer> numbers = numberGenerator.generateRandomNumber();
+            LottoTicket ticket = new LottoTicket(numbers);
+            tickets.add(ticket);
+        }
+        return tickets;
+    }
```
여기서 return 해서 밖에서 사용하면 캡슐화가 깨질 것 같아요!

생성자에 ticketCount와 NumberGenerator를 받고 생성자에서 만든 후 필드로 값을 넣어주는 건 어떨까요?

즉 LottoTickets라는 일급컬렉션은 List<Integer> 필드 하나만 가지고 생성자에서 만들고, 필요하다면 밖에서 사용하는 식으로요!

sosow0212 commented on src/main/java/domain/RandomNumberGenerator.java:
```diff
@@ -0,0 +1,29 @@
+    private static final Random random = new Random();
+
+    @Override
+    public List<Integer> generateRandomNumber() {
+        List<Integer> numbers = new ArrayList<>();
+        while (numbers.size() < LOTTO_SIZE) {
+            checkDuplicateNumber(numbers);
+        }
+        return numbers;
+    }
+
+    private static void checkDuplicateNumber(List<Integer> number) {
```
static일 필요가 없을 것 같습니다!

sosow0212 commented on src/main/java/domain/TicketCount.java:
```diff
@@ -0,0 +1,16 @@
+package domain;
+
+public class TicketCount {
+
+    private static final int TICKET_PRICE = 1000;
+
+    private final int money;
+
+    public TicketCount(int money) {
+        this.money = money;
+    }
+
+    public int getCount() {
+        return money / TICKET_PRICE;
+    }
```
이 부분도 TIcketCount라는 네이밍과 걸맞게 필드를 `int count`로 두고 생성자에서 money를 받고 필드 값을 채워주는 방식으로 하고 추후에 getCount()로 필드만 꺼내주게 해도 좋을 것 같아요~

sosow0212 commented on src/main/java/domain/WinningLotto.java:
```diff
@@ -0,0 +1,29 @@
+
+import java.util.List;
+
+public class WinningLotto {
```
이 방법이 잘못된 건 아니지만 우승 로또를 찾아주는 것이라면 일급컬렉션 `LottoTickets`에서 찾아주는 게 좋을 것 같아요!
현재 구조라면 controller에서 요소 하나하나를 꺼내서 해당 객체에 넘겨줄 것 같아요.
따라서 현재 방식은 응집도가 떨어지기도 하고 다른 클래스의 컬렉션 값을 외부 클래스에서 마구 바꿀 수 있다는 단점도 있을 것 같습니다!

이를 개선하기 위해서 일급컬렉션에 나오는 개념을 지키기 위해서 `LottoTickets`의 필드가 `List<LottoTicket> lottos`만 있어야하고, 현재 로직은 일급 컬렉션으로 위임을하면 객체에게 걸맞는 일을 시킬 수 있을 것 같네요!

sosow0212 reviewed (approved):
코드도 예뻐졌고 구조적으로 생각할 수 있게 된 것 같아서 너무 뿌듯합니다! 이 맛에 리뷰하는 것 같아요.
객체지향의 사실과오해 책이 효과가 있었나요 😎

아직은 조금 아쉬운 부분이 있긴 하지만, 객체에게 일을 시키려고 하는 모습이 보여서 너무 뿌듯하기도 하고 이전보다 더 성장한 것 같습니다. 아쉬운 부분이야 앞으로 채워나가면 되기 때문에 너무 걱정하진 마시고 지금처럼만 하시면 좋을 것 같아요.

일단 이번 단계는 여기서 Merge하겠습니다.
추가로 남긴 리뷰 확인해보시고 고민해보셨으면 좋겠습니다. 그리고 계속 `No newline at end of file(EOF)` 문제가 생기는데 이것도 해결해주시고, 누락된 테스트도 다음 단계에서 같이 작성하고 제출해주세요!

수고 많으셨습니다 👏
